### PR TITLE
[experiment] strongly typed events with polysemy

### PR DESCRIPTION
**Pull request by fisx**
I don't expect this PR to go anywhere, just wanted to learn something.  Issue with findings coming up soon!

### CODE REVIEW

File: services/brig/src/Brig/IO/Intra.hs [hunk 1]
```diff
@@ -94,17 +58,44 @@ import Imports
...
+
 -----------------------------------------------------------------------------
 -- Event Handlers
 
-onUserEvent :: UserId -> Maybe ConnId -> UserEvent -> AppIO ()
-onUserEvent orig conn e =
+onUserEvent :: IsUserEvent e => UserId -> Maybe ConnId -> e -> AppIO ()
+onUserEvent orig conn e = do
+  -- I am tempted to try and use 'Chan' to simply drop events, and then let updateSearchIndex,
+  -- dispatchNotifications, and journalEvent all subscribe to duplicates of that 'Chan'.  This
```
Comment (fisx) [inline services/brig/src/Brig/IO/Intra.hs:93]: or maybe http://hackage.haskell.org/package/broadcast-chan

File: services/brig/src/Brig/IO/Intra.hs [hunk 2]
```diff
@@ -154,73 +145,32 @@ onClientEvent orig conn e = do
...
-    Journal.userUpdate orig Nothing (Just loc) Nothing
-  UserIdentityUpdated (UserIdentityUpdatedData _ (Just em) _) ->
-    Journal.userUpdate orig (Just em) Nothing Nothing
-  UserIdentityRemoved (UserIdentityRemovedData _ (Just em) _) ->
-    Journal.userEmailRemove orig em
-  UserDeleted {} ->
-    Journal.userDelete orig
-  _ ->
-    return ()
+-- (this function shows that things get a lot simpler and type info increases, but i think we
```
Comment (mheinzel) [inline services/brig/src/Brig/IO/Intra.hs:148]: > this function shows that things get a lot simpler and type info increases

not sure what exactly you mean. What got simpler and where is the additional type info?

**Comment by arianvp** [comment]
I don't think this gains us a lot. We still have to manually trigger the events in the code. which is ad hoc.

My problem with our events code is not that we dont track it in the type, it's that we manually have to call it all the time; and if you forget you get bugs. Instead event emissions should be a pattern that we abstract.  They should reflect exactly the operations the database is doing. The only thing that is different is that not all state is synced to all users; and that's the only thing we should differentiate on.

**Comment by arianvp** [comment]
Instead of describing in the type what events we can fire by calling callbacks as _input_

functions should return what events they emit as _output_. This seems a way more natural way to express the problem

**Comment by fisx** [comment]
> Instead of describing in the type what events we can fire by calling callbacks as _input_
> 
> functions should return what events they emit as _output_. This seems a way more natural way to express the problem

I don't think we're looking at the same code, because I agree with you, and I think that's what I've outlined in this PR.  :)  I'll try to get on the phone with you about this on Friday.

File: services/brig/src/Brig/IO/Intra.hs [hunk 3]
```diff
@@ -94,17 +58,44 @@ import Imports
 import Network.HTTP.Types.Method
 import Network.HTTP.Types.Status
 import qualified Network.Wai.Utilities.Error as Wai
+import Polysemy
 import System.Logger.Class as Log hiding (name, (.=))
 import Wire.API.Team.Feature (TeamFeatureName (..), TeamFeatureStatus)
 
+----------------------------------------------------------------------
+-- polysemi
```
Comment (akshaymankar) [inline services/brig/src/Brig/IO/Intra.hs:66]:
```suggestion
-- polysemy
```

**Comment by fisx** [comment]
this is not going to get merged, but i've linked it here [here](https://github.com/zinfra/backend-issues/issues/1846).

**Pull request closed without merging**